app/src/org/commcare/android/nfc/NfcActivity.java

```diff
@@ -0,0 +1,150 @@
+package org.commcare.android.nfc;
+
+import android.app.Activity;
+import android.app.PendingIntent;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.nfc.NfcAdapter;
+import android.nfc.Tag;
+import android.os.Build;
+import android.os.Bundle;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.logging.ReportingUtils;
+import org.commcare.dalvik.R;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * Parent activity that provides all of the functionality common to any NFC action that CommCare
+ * supports
+ *
+ * @author Aliza Stone
+ */
+public abstract class NfcActivity extends Activity {
+
+    protected static final String NFC_PAYLOAD_MULT_TYPES_ARG = "types";
+    protected static final String NFC_PAYLOAD_SINGLE_TYPE_ARG = "type";
+    protected static final String NFC_DOMAIN_ARG = "domain";
+
+    protected NfcManager nfcManager;
+    protected PendingIntent pendingNfcIntent;
+    protected String domainForType;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
+            initFields();
+            createPendingRestartIntent();
+            setContentView(R.layout.nfc_instructions_view);
+            ((TextView)findViewById(R.id.nfc_instructions_text_view)).
+                    setText(Localization.get(getInstructionsTextKey()));
+        }
+    }
+
+    protected void initFields() {
+        this.nfcManager = new NfcManager(this);
+
+        this.domainForType = getIntent().getStringExtra(NFC_DOMAIN_ARG);
+        if (this.domainForType == null) {
+            // Usually there is no reason for the user to provide this themselves, and we will just
```

Comment:
Mm this may be a little bit YAGNI, but my thinking was that if any app ever needs to read a value that was originally written by an app from a different domain, then it would be important to have this override option.